### ISSUE TITLE
ES regression disabling workers sporadically

### ISSUE THREAD

**Issue by wil93** [cms-dev/cms#661]
Sometimes I get this error:

```
[...]
2016-10-01 11:33:58,206 - WARNING [Evaluation,0] Integrity error while inserting worker result.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/cms-1.3.dev0-py2.7.egg/cms/service/EvaluationService.py", line 645, in write_results_one_object_and_type
    session, object_result, operation, result)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 483, in __exit__
    self.rollback()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/langhelpers.py", line 60, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 480, in __exit__
    self.commit()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 385, in commit
    self._prepare_impl()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 365, in _prepare_impl
    self.session.flush()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 1986, in flush
    self._flush(objects)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 2104, in _flush
    transaction.rollback(_capture_exception=True)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/langhelpers.py", line 60, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 2068, in _flush
    flush_context.execute()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/unitofwork.py", line 372, in execute
    rec.execute(self)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/unitofwork.py", line 526, in execute
    uow
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/persistence.py", line 66, in save_obj
    mapper, table, insert)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/persistence.py", line 603, in _emit_insert_statements
    execute(statement, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 841, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/sql/elements.py", line 322, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 938, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1070, in _execute_context
    context)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1271, in _handle_dbapi_exception
    exc_info
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/compat.py", line 199, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
IntegrityError: (IntegrityError) duplicate key value violates unique constraint "evaluations_submission_id_dataset_id_testcase_id_key"
DETAIL:  Key (submission_id, dataset_id, testcase_id)=(95219, 229, 6022) already exists.
 'INSERT INTO evaluations (submission_id, dataset_id, testcase_id, outcome, text, execution_time, execution_wall_clock_time, execution_memory, evaluation_shard, evaluation_sandbox) VALUES (%(submission_id)s, %(dataset_id)s, %(testcase_id)s, %(outcome)s, %(text)s, %(execution_time)s, %(execution_wall_clock_time)s, %(execution_memory)s, %(evaluation_shard)s, %(evaluation_sandbox)s) RETURNING evaluations.id' {'execution_memory': 262144, 'submission_id': 95219, 'testcase_id': 6022, 'execution_time': 0.003, 'text': '["Execution failed because the return code was nonzero"]', 'evaluation_shard': 0, 'dataset_id': 229, 'outcome': u'0.0', 'evaluation_sandbox': u'/tmp/tmpbGmwRe/tmp', 'execution_wall_clock_time': 0.086}
[...]
```

Sometimes, this:

```
[...]
2016-10-01 15:57:15,376 - WARNING [Evaluation,0] Integrity error while inserting worker result.
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/cms-1.3.dev0-py2.7.egg/cms/service/EvaluationService.py", line 645, in write_results_one_object_and_type
    session, object_result, operation, result)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 483, in __exit__
    self.rollback()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/langhelpers.py", line 60, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 480, in __exit__
    self.commit()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 385, in commit
    self._prepare_impl()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 365, in _prepare_impl
    self.session.flush()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 1986, in flush
    self._flush(objects)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 2104, in _flush
    transaction.rollback(_capture_exception=True)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/langhelpers.py", line 60, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/session.py", line 2068, in _flush
    flush_context.execute()
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/unitofwork.py", line 372, in execute
    rec.execute(self)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/unitofwork.py", line 526, in execute
    uow
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/persistence.py", line 66, in save_obj
    mapper, table, insert)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/orm/persistence.py", line 603, in _emit_insert_statements
    execute(statement, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 841, in execute
    return meth(self, multiparams, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/sql/elements.py", line 322, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 938, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1070, in _execute_context
    context)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1271, in _handle_dbapi_exception
    exc_info
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/util/compat.py", line 199, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/base.py", line 1063, in _execute_context
    context)
  File "/usr/local/lib/python2.7/dist-packages/sqlalchemy/engine/default.py", line 442, in do_execute
    cursor.execute(statement, parameters)
IntegrityError: (IntegrityError) duplicate key value violates unique constraint "executables_submission_id_dataset_id_filename_key"
DETAIL:  Key (submission_id, dataset_id, filename)=(95275, 312, capslock) already exists.
 'INSERT INTO executables (submission_id, dataset_id, filename, digest) VALUES (%(submission_id)s, %(dataset_id)s, %(filename)s, %(digest)s) RETURNING executables.id' {'submission_id': 95275, 'dataset_id': 312, 'digest': u'34ad0a61cd22b7d967aeaf64c10f3a26bff6c68f', 'filename': u'capslock'}
[...]
```

And then, sporadically, this happens:

```
[...]
2016-10-01 16:12:37,133 - INFO [Evaluation,0] Executing operations `evaluate on 95278 against dataset 317, testcase 015' and 2 more.
2016-10-01 16:12:38,458 - INFO [Evaluation,0] `evaluate on 95278 against dataset 317, testcase 012' completed. Success: True.
2016-10-01 16:12:38,459 - INFO [Evaluation,0] `evaluate on 95278 against dataset 317, testcase 013' completed. Success: True.
2016-10-01 16:12:38,462 - INFO [Evaluation,0] `evaluate on 95278 against dataset 317, testcase 014' completed. Success: True.
2016-10-01 16:12:38,466 - INFO [Evaluation,0] Asking worker 0 to `evaluate on 95278 against dataset 317, testcase 015'.
2016-10-01 16:12:38,487 - INFO [Evaluation,0] Asking worker 0 to `evaluate on 95278 against dataset 317, testcase 016'.
2016-10-01 16:12:38,501 - INFO [Evaluation,0] Asking worker 0 to `evaluate on 95278 against dataset 317, testcase 017'.
2016-10-01 16:12:38,513 - INFO [Evaluation,0] Operations `evaluate on 95278 against dataset 317, testcase 015' and 2 more concluded successfully.
2016-10-01 16:12:38,514 - INFO [Evaluation,0] Ratio is 2, executing 2 operations together.
2016-10-01 16:12:38,515 - INFO [Evaluation,0] Executing operations `evaluate on 95278 against dataset 317, testcase 018' and 1 more.
2016-10-01 16:12:39,867 - INFO [Evaluation,0] Start looking for missing operations.
2016-10-01 16:12:40,903 - INFO [Evaluation,0] Found 3 missed operation(s) in 1035 ms.
2016-10-01 16:12:40,905 - INFO [Evaluation,0] `evaluate on 95278 against dataset 317, testcase 015' completed. Success: True.
2016-10-01 16:12:40,912 - INFO [Evaluation,0] Asking worker 0 to `evaluate on 95278 against dataset 317, testcase 018'.
2016-10-01 16:12:40,930 - INFO [Evaluation,0] Asking worker 0 to `evaluate on 95278 against dataset 317, testcase 019'.
2016-10-01 16:12:40,945 - INFO [Evaluation,0] Operations `evaluate on 95278 against dataset 317, testcase 018' and 1 more concluded successfully.
2016-10-01 16:12:40,946 - INFO [Evaluation,0] Ratio is 2, executing 2 operations together.
2016-10-01 16:12:40,947 - INFO [Evaluation,0] Executing operations `evaluate on 95278 against dataset 317, testcase 003' and 1 more.
2016-10-01 16:14:36,867 - INFO [Evaluation,0] Start looking for missing operations.
2016-10-01 16:24:45,579 - ERROR [Evaluation,0] Disabling and shutting down worker 0 because of no response in 0:12:04.673155.
2016-10-01 16:24:45,581 - INFO [Evaluation,0] Worker 0 released and disabled.
2016-10-01 16:24:45,582 - ERROR [Evaluation,0] Disabling and shutting down worker 1 because of no response in 0:12:08.550726.
2016-10-01 16:24:45,582 - INFO [Evaluation,0] Worker 1 released and disabled.
2016-10-01 16:24:45,583 - INFO [Evaluation,0] Operation evaluate on 95278 against dataset 317, testcase 018 put again in the queue because of worker timeout.
2016-10-01 16:24:45,943 - INFO [Evaluation,0] Terminated connection with localhost:26001 (Worker,1): Connection closed.
2016-10-01 16:24:45,945 - INFO [Evaluation,0] Terminated connection with localhost:26000 (Worker,0): Connection closed.
2016-10-01 16:24:49,963 - INFO [Evaluation,0] Established connection with localhost:26001 (Worker,1).
2016-10-01 16:24:49,964 - INFO [Evaluation,0] Worker 1 online again.
2016-10-01 16:24:49,965 - INFO [Evaluation,0] Established connection with localhost:26000 (Worker,0).
2016-10-01 16:24:49,965 - INFO [Evaluation,0] Worker 0 online again.
[...]
```

^ at this point, there is no more log data for _hours_, until I manually restarted everything.

**Comment by lucach** [comment]
This should be caused by race conditions in WorkerPool (commit 4757518).

**Comment by wil93** [comment]
This was fixed in #664
